fix(GraphQL) Refs #29270  Improve error message for unauthorized page request

valentinogiardino opened the pull request:
### Proposed Changes
Enhance the error handling capabilities of our GraphQL API by introducing a custom exception class that aligns with the GraphQL error handling guidelines.
* Added a custom GraphQL exception class that implements `GraphQLError`.
* Introduced `PermissionDeniedGraphQLException` for unauthorized access.
* Fixed error message for unauthorized page requests.
* Added Postman test for exception handling.

### Checklist
- [x] Unauthorized query request should return a correct error message

### Screenshots
#### Before:
![image](https://github.com/user-attachments/assets/8ee2cd67-af11-4616-abec-534dd2b56b88)
#### After:
![image](https://github.com/user-attachments/assets/8dd74fb9-d0ee-4c80-8d99-6056ade8eac5)

This PR fixes #29270

This PR fixes: #29270

valentinogiardino committed:
#29270 add graphql custom exception class

valentinogiardino committed:
#29270 add PermissionDeniedGraphQLException

valentinogiardino committed:
#29270 fix error message for Unauthorized Page Request

valentinogiardino committed:
#29270 add postman test

fabrizzio-dotCMS commented on dotCMS/src/main/java/com/dotcms/graphql/exception/CustomGraphQLException.java:
```diff
@@ -0,0 +1,39 @@
+        this.errorCode = errorCode;
+    }
+
+    @Override
```
Is the front end prepared to deal with this "extensions" attribute?
I mean is there any particular reason we want to have it ?

fabrizzio-dotCMS commented on dotCMS/src/main/java/com/dotcms/graphql/exception/CustomGraphQLException.java:
```diff
@@ -0,0 +1,39 @@
+package com.dotcms.graphql.exception;
```
I think we could simply throw `GraphQLException` here and pass one or the other message instead

fabrizzio-dotCMS replied:
yeah I informed myself a little bit better and now I can see why you did what you did here.

fabrizzio-dotCMS commented on dotCMS/src/main/java/com/dotcms/graphql/exception/CustomGraphQLException.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
```
If we had to use this `CustomException` perhaps  `GraphQLException` is more suited for Base Class

fabrizzio-dotCMS commented on dotCMS/src/main/java/com/dotcms/graphql/exception/CustomGraphQLException.java:
```diff
@@ -0,0 +1,39 @@
+        customAttributes.put("errorCode", this.errorCode);
+        customAttributes.put("errorMessage", this.getMessage());
+
+        return customAttributes;
```
if we need to keep this.. of which I am not certain. 
The Pattern would be this: 
`return Map.of("errorCode",,this.errorCode,"errorMessage", this.getMessage());`
Try to use immutable class to return results as mush as possible

fabrizzio-dotCMS replied:
![Screenshot 2024-08-09 at 3 36 22 PM](https://github.com/user-attachments/assets/bc2bb166-444f-47b1-8d8e-fe42f6cb7569)

The code returned by the endpoint is still 200, this code needs to be changed to match the permissions issue
We have exceptions to code mapping configured in `ReloadableServletContainer` Check it out, Perhaps all you need to do is use one of our own exceptions passing the right message and that's it.

valentinogiardino replied:
As far as I know, the HTTP status code returned is typically 200, regardless of whether the response contains data or errors. This is because the HTTP status code reflects the success of the HTTP request itself, not the success of the GraphQL query. While customizing status codes is possible, it’s not standard practice as it goes against GraphQL’s principle of being transport-agnostic.

If we still want the status code to match the specific error thrown by the server, we would need to implement middleware that inspects all error responses, categorizes each error based on its type, and sets the corresponding status code. However, this could become more complex if the response contains multiple errors. Given the scope of this issue, I believe this approach is beyond what is currently needed.

fabrizzio-dotCMS replied:
you're right about this.

valentinogiardino committed:
#29270 use GraphqlErrorException as base class for custom graphql exceptions

valentinogiardino committed:
#29270 fix postman test

valentinogiardino committed:
Merge branch 'master' into 29270-graphql-improve-error-message-for-unauthorized-page-request

valentinogiardino committed:
#29270 Rollback to `CustomGraphQLException` as Base Class

**Reason for Rollback**:
- The change to use `GraphQLErrorException` as the base class for `PermissionDeniedGraphQLException` caused a violation of the SonarQube rule "Inheritance tree of classes should not be too deep."
- The inheritance tree depth exceeded the allowed limit, leading to maintainability issues flagged by SonarQube.

valentinogiardino committed:
Merge branch 'master' into 29270-graphql-improve-error-message-for-unauthorized-page-request

valentinogiardino committed:
Merge branch 'master' into 29270-graphql-improve-error-message-for-unauthorized-page-request
